File: interview/algorithms/sortings/counting_sort.py
import os
import sys

# assume all elements in arr are <= len(arr)
def counting_sort(arr):
    # maintain a count of each element in the array
    aux_arr = [0 for _ in range(len(arr)+1)] # len(arr) == 8, len(aux_arr) = 9
    for i in range(len(arr)):
        aux_arr[arr[i]] += 1

    # aux_arr[index] contains the number of times index is present in arr

    # modify aux_arr so that value of each index contains the nunber of items equal to index or less than that
    for i in range(1, len(aux_arr)):
        aux_arr[i] += aux_arr[i-1]

    # print aux arr for debugging
    print('aux_arr {}'.format(aux_arr))

    # create the target arr now
    target_arr = [None] * len(arr)
    for i in range(len(arr)-1, -1, -1):
        aux_arr[arr[i]] -= 1
        target_arr[aux_arr[arr[i]]] = arr[i]
    return target_arr


if __name__ == "__main__":
    # prepare input
    arr = [2, 5, 3, 0, 2, 3, 8, 3]
    print('un-sorted array {}'.format(arr))
    arr = counting_sort(arr)
    print('sorted array {}'.format(arr))

File: interview/algorithms/sortings/counting_sort.cpp
// counting sorting. Linear order sorting used to sort elements n elements all 
// of which are in the range 0..k
#include <iostream>
#include <algorithm>
#include <random>
#include <iterator>
#include <limits>
#include <iomanip>
#include <cstring>
#include <utility>
#include <ctime>
using namespace std;

template <size_t N>
void counting_sort(int (&arr)[N], int k) {
    // first create an array of size k+1
    int c[k+1] {};

    // count how many times each element occurs in the input arr
    for (int elem : arr) {
        ++c[elem];
    }

    // count the number of elements equal or less than the current element
    for (int idx = 1; idx <= k; ++idx) {
        c[idx] += c[idx-1];
    }

    // make a copy of the input array and then modify it.
    int arr_copy[N] {};
    std::copy(begin(arr), end(arr), begin(arr_copy));
    for(int elem : arr_copy) {
        arr[c[elem]-1] = elem;
        --c[elem];
    }
}

int main() {
    // n = 10
    // k = 0..15, so fill arr with random elements between 0 and 15 (both inclusive)
    int arr[10] {};

    std::random_device r;
    std::default_random_engine engine(r());
    std::uniform_int_distribution<int> dist(0, 15);
    auto gen = [dist = dist, engine = engine] () mutable {
        return dist(engine);
    };
    std::generate(begin(arr), end(arr), gen);

    cout << "Elements before\n";
    std::copy(begin(arr), end(arr), ostream_iterator<int>(cout, " "));
    counting_sort(arr, 15);
    cout << "\nElements sorted\n";
    std::copy(begin(arr), end(arr), ostream_iterator<int>(cout, " "));
    cout << endl;
}